documentation/examples/tail-wal/main.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"os"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/wal"
+	"gopkg.in/alecthomas/kingpin.v2"
+	fsnotify "gopkg.in/fsnotify/fsnotify.v1"
+)
+
+var (
+	a      = kingpin.New("WAL tail example", "Example program for tailing a Prometheus WAL.")
+	walDir = a.Arg("wal.dir", "Full path to a Prometheus WAL directory.").Required().String()
+	logger = log.With(log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr)), "ts", log.DefaultTimestampUTC, "caller", log.DefaultCaller)
+)
+
+func decode(segment *wal.Segment) bool {
+	// todo: callum, is there an easy way to detect if r.Next() has returned the last possible record in a segment.
+	// For now just recover from the panic that is thrown if we call r.Next() when we shouldn't have
+	defer func() {
+		if r := recover(); r != nil {
+			level.Error(log.With(logger)).Log("err", r)
+			segment.Close()
+		}
+	}()
+	r := wal.NewReader(segment)
+	var (
+		dec      tsdb.RecordDecoder
+		series   []tsdb.RefSeries
+		samples  []tsdb.RefSample
+		nSeries  = 0
+		nSamples = 0
+	)
+	for {
+		ok := r.Next()
+		rec := r.Record()
+		switch dec.Type(rec) {
+		case tsdb.RecordSeries:
+			series, err := dec.Series(rec, series[:0])
+			if err != nil {
+				level.Error(log.With(logger)).Log("err", err)
+				return ok
+			}
+			nSeries += len(series)
+		case tsdb.RecordSamples:
+			samples, err := dec.Samples(rec, samples[:0])
+			if err != nil {
+				level.Error(log.With(logger)).Log("err", err)
+				return ok
+			}
+			nSamples += len(samples)
+		}
+		if !ok {
+			return ok
+		}
+	}
+}
+
+func main() {
+	a.HelpFlag.Short('h')
+
+	_, err := a.Parse(os.Args[1:])
+	if err != nil {
+		level.Error(log.With(logger)).Log("err", err)
+		return
+	}
+	// change this to your wal dir
+	// dir := "/home/callum/go/src/github.com/prometheus/prometheus/data/wal/"
+	w, err := wal.New(nil, nil, *walDir)
+	if err != nil {
+		level.Error(log.With(logger)).Log("err", err)
+	}
+	m, n, err := w.Segments()
+
+	if err != nil {
+		level.Error(log.With(logger)).Log("err", err)
+	}
+	if n == -1 {
+		level.Error(log.With(logger)).Log("err", "no segments found")
+		return
+	}
+
+	// Read the existing segments except the last one.
+	var segment *wal.Segment
+	for ; m < n; m++ {
+		segment, err = wal.OpenReadSegment(wal.SegmentName(*walDir, m))
+		if err != nil {
+			level.Error(log.With(logger)).Log("err", err)
+		}
+		defer segment.Close()
+		for {
```

Comment:
This could just be `for decode(segment) {}`, though even better would be perhaps decode only returns when it is done reading the entire segment regardless of a decoding error? Or else errors need to be handled somewhere?